src/ram_key_value_store/mod.rs

```diff
@@ -0,0 +1,128 @@
+use super::event::Event;
+use super::key_value_store::Key;
+use super::key_value_store::KeyValueStore;
+use super::key_value_store::Size;
+use super::key_value_store::Storable;
+use core::cell::Cell;
+use core::mem;
+
+#[cfg(test)]
+mod test;
+
+pub struct RamKeyValueStoreElement {
+    pub key: Key,
+    pub size: Size,
+}
+
+impl RamKeyValueStoreElement {
+    pub fn new<T: Sized>(key: Key) -> Self {
+        Self {
+            key: key,
+            size: mem::size_of::<T>() as u8,
+        }
+    }
+}
+
+pub struct RamKeyValueStore<'a> {
+    ram: &'a mut [Cell<u8>],
+    elements: &'a [RamKeyValueStoreElement],
+    on_change_event: Event<'a, Key>,
+}
+
+impl<'a> RamKeyValueStore<'a> {
+    pub fn new(ram: &'a mut [Cell<u8>], elements: &'a [RamKeyValueStoreElement]) -> Self {
+        let required_size: usize = elements.iter().map(|x| x.size as usize).sum();
+        assert!(ram.len() == required_size, "Incorrect RAM size");
+
+        Self {
+            ram,
+            elements,
+            on_change_event: Event::<'a, Key>::new(),
+        }
+    }
+}
+
+impl<'a> KeyValueStore<'a> for RamKeyValueStore<'a> {
+    fn read(&self, key: Key, value: &mut dyn Storable) {
+        let mut offset = 0 as usize;
+
+        for element in self.elements.iter() {
+            if key == element.key {
+                assert!(value.size() == element.size, "Invalid size");
+                assert!(
+                    value.can_deserialize_from(&self.ram[offset..offset + element.size as usize]),
```

Comment:
I still think this looks kinda silly since to me since `can_deserialize_from` _should_ validate the given slice, and line 51 just repeats what I would think is an integral part of said validation 😅

Reply:
I don't disagree... but why do it in every implementation instead of just doing it once here? The extra code matters in an embedded context.

Plus it would be equally weird for this to defer to the object in `read` while doing the size check in `write`.

IMO, any solution is going to be a compromise.